src/systems/systems/src/pressurization/cabin_pressure_controller.rs

```diff
@@ -0,0 +1,455 @@
+        const INTERNAL_VS_RATE_CHANGE: f64 = 100.; // Rate of change of 100fpm per second
+
+        let rate_of_change_for_delta = INTERNAL_VS_RATE_CHANGE * context.delta_as_secs_f64();
+        if self.cabin_target_vs > self.cabin_vs() {
+            self.cabin_vs
+                + Velocity::new::<foot_per_minute>(rate_of_change_for_delta.min(
+                    self.cabin_target_vs.get::<foot_per_minute>()
+                        - self.cabin_vs().get::<foot_per_minute>(),
+                ))
+        } else if self.cabin_target_vs < self.cabin_vs() {
+            self.cabin_vs
+                - Velocity::new::<foot_per_minute>(rate_of_change_for_delta.min(
+                    self.cabin_vs().get::<foot_per_minute>()
+                        - self.cabin_target_vs.get::<foot_per_minute>(),
+                ))
+        } else {
+            self.cabin_target_vs
+        }
+    }
+
+    pub fn cabin_pressure(&self) -> Pressure {
+        self.cabin_pressure
+    }
+
+    pub fn cabin_altitude(&self) -> Length {
+        self.cabin_alt
+    }
+
+    pub fn cabin_vs(&self) -> Velocity {
+        self.cabin_vs
+    }
+
+    pub fn cabin_delta_p(&self) -> Pressure {
+        self.cabin_pressure - self.exterior_pressure
+    }
+
+    pub fn pressure_schedule(&self) -> PressureSchedule {
+        self.pressure_schedule_manager.pressure_schedule()
+    }
+
+    pub fn should_switch_cpc(&self) -> bool {
+        self.pressure_schedule_manager.should_switch_cpc()
+    }
+
+    pub fn reset_cpc_switch(&mut self) {
+        self.pressure_schedule_manager.reset_cpc_switch();
+    }
+}
+
+impl PressureValveActuator for CabinPressureController {
+    fn should_open_pressure_valve(&self) -> bool {
+        self.cabin_vs < Velocity::new::<meter_per_second>(0.)
+            || self.pressure_schedule() == PressureSchedule::Ground
+    }
+
+    fn should_close_pressure_valve(&self) -> bool {
+        self.cabin_vs > Velocity::new::<meter_per_second>(0.)
+    }
+
+    fn target_valve_position(&self) -> Ratio {
+        // Placeholder formula to simulate OFV behaviour based on empirical data
+        let pressure_ratio: f64 = self.exterior_pressure.get::<hectopascal>()
+            / (self.cabin_pressure().get::<hectopascal>() * 2.);
+        let vs_ratio: f64 = self.cabin_target_vs.get::<foot_per_minute>() / 2000. * pressure_ratio;
+        if (pressure_ratio + vs_ratio) > 1.
+            || (self.pressure_schedule() == PressureSchedule::Ground
+                && self.pressure_schedule_manager.should_open_outflow_valve())
+        {
+            Ratio::new::<percent>(100.)
+        } else if (pressure_ratio + vs_ratio) < 0. {
+            Ratio::new::<percent>(0.)
+        } else {
+            Ratio::new::<percent>((pressure_ratio + vs_ratio) * 100.)
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum PressureSchedule {
+    Ground,
+    TakeOff,
+    ClimbInternal,
+    Cruise,
+    DescentInternal,
+    Abort,
+}
```

Comment:
**question**: Why are Climb and Descent suffixed with Internal?

Reply:
This comes straight from the Airbus' AMM documentation. I have no idea why these two are suffixed Internal and the others aren't, but I thought I'd keep it like this for accuracy with the real system.

![image](https://user-images.githubusercontent.com/60938089/125170216-c96ab880-e1a5-11eb-89e0-8664d34e589a.png)

Reply:
Interesting. Makes sense then. Curious why they added "internal".